spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/core/Mqttv3ClientManager.java

```diff
@@ -0,0 +1,191 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mqtt.core;
+
+import java.time.Instant;
+import java.util.concurrent.ScheduledFuture;
+
+import org.eclipse.paho.client.mqttv3.IMqttAsyncClient;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.MqttCallback;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.beans.factory.BeanFactoryAware;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.integration.context.IntegrationContextUtils;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.Assert;
+
+public class Mqttv3ClientManager extends AbstractMqttClientManager<IMqttAsyncClient>
+		implements MqttCallback, InitializingBean, BeanFactoryAware {
+
+	/**
+	 * The default reconnect timeout in millis.
+	 */
+	private static final long DEFAULT_RECOVERY_INTERVAL = 10_000;
+
+	private final MqttPahoClientFactory clientFactory;
+
+	private BeanFactory beanFactory;
+
+	private TaskScheduler taskScheduler;
+
+	private volatile ScheduledFuture<?> scheduledReconnect;
+
+	private volatile IMqttAsyncClient client;
+
+	private long recoveryInterval = DEFAULT_RECOVERY_INTERVAL;
+
+	public Mqttv3ClientManager(MqttPahoClientFactory clientFactory, String clientId) {
+		super(clientId);
+		Assert.notNull(clientFactory, "'clientFactory' is required");
+		this.clientFactory = clientFactory;
+		String[] serverURIs = clientFactory.getConnectionOptions().getServerURIs();
+		Assert.notEmpty(serverURIs, "'serverURIs' must be provided in the 'MqttConnectionOptions'");
+		setUrl(serverURIs[0]);
+	}
+
+	public Mqttv3ClientManager(String url, String clientId) {
+		super(clientId);
+		Assert.notNull(url, "'url' is required");
+		setUrl(url);
+		MqttConnectOptions connectOptions = new MqttConnectOptions();
+		connectOptions.setServerURIs(new String[]{ url });
+		DefaultMqttPahoClientFactory defaultFactory = new DefaultMqttPahoClientFactory();
+		defaultFactory.setConnectionOptions(connectOptions);
+		this.clientFactory = defaultFactory;
+	}
+
+	@Override
+	public IMqttAsyncClient getClient() {
+		return this.client;
+	}
+
+	@Override
+	public void afterPropertiesSet() {
+		this.taskScheduler = IntegrationContextUtils.getTaskScheduler(this.beanFactory);
```

Comment:
You know I got another thought: how about to take an opportunity and make this v3 client manager logically similar to what we have so far for v5?
Kinda no manual reconnect, but that `this.connectionOptions.isAutomaticReconnect()`.
See what I did in the `Mqttv5PahoMessageDrivenChannelAdapter` for reconnections: it does not use a `taskScheduler`.
There is just that logic in the `doStart()` impl.
I believe the same algorithm will work for v3, too.

Reply:
Yeah, that was my consideration too. I forgot to ask you about it though, if you think there was any particular reason not to use v3 client re-connection rather than do it manually 🙂